source/auth/auth.md

```diff
@@ -0,0 +1,2004 @@
+
+#### Step 3
+
+For test users that support only one mechanism, verify that explicitly specifying the other mechanism fails.
+
+For a non-existent username, verify that not specifying a mechanism when connecting fails with the same error type that
+would occur with a correct username but incorrect password or mechanism. (Because negotiation with a non-existent user
+name at one point during server development caused a handshake error, we want to verify this is seen by users as similar
+to other authentication errors, not as a network or database command error on the `hello` or legacy hello commands
+themselves.)
+
+#### Step 4
+
+To test SASLprep behavior, create two users:
+
+1. username: "IX", password "IX"
+1. username: "\\u2168" (ROMAN NUMERAL NINE), password "\\u2163" (ROMAN NUMERAL FOUR)
+
+To create the users, use the exact bytes for username and password without SASLprep or other normalization and specify
+SCRAM-SHA-256 credentials:
+
+```
+db.runCommand({createUser: 'IX', pwd: 'IX', roles: ['root'], mechanisms: ['SCRAM-SHA-256']})
+db.runCommand({createUser: '\\u2168', pwd: '\\u2163', roles: ['root'], mechanisms: ['SCRAM-SHA-256']})
+```
+
+For each user, verify that the driver can authenticate with the password in both SASLprep normalized and non-normalized
+forms:
+
+- User "IX": use password forms "IX" and "I\\u00ADX"
+- User "\\u2168": use password forms "IV" and "I\\u00ADV"
+
+As a URI, those have to be UTF-8 encoded and URL-escaped, e.g.:
+
+- <mongodb://IX:IX@mongodb.example.com/admin>
+- <mongodb://IX:I%C2%ADX@mongodb.example.com/admin>
+- <mongodb://%E2%85%A8:IV@mongodb.example.com/admin>
+- <mongodb://%E2%85%A8:I%C2%ADV@mongodb.example.com/admin>
+
+### Speculative Authentication
+
+See the speculative authentication section in the
+[MongoDB Handshake spec](https://github.com/mongodb/specifications/blob/master/source/mongodb-handshake/handshake.rst).
+
+### Minimum iteration count
+
+For SCRAM-SHA-1 and SCRAM-SHA-256, test that the minimum iteration count is respected. This may be done via unit testing
+of an underlying SCRAM library.
+
+## Backwards Compatibility
+
+Drivers may need to remove support for association of more than one credential with a MongoClient, including
+
+- Deprecation and removal of MongoClient constructors that take as an argument more than a single credential
+- Deprecation and removal of methods that allow lazy authentication (i.e post-MongoClient construction)
+
+Drivers need to support both the shorter and longer SCRAM-SHA-1 and SCRAM-SHA-256 conversations over MongoDB's SASL
+implementation. Earlier versions of the server required an extra round trip due to an implementation decision. This was
+accomplished by sending no bytes back to the server, as seen in the following conversation (extra round trip
+emphasized):
+
+C:
+`{saslStart: 1, mechanism: "SCRAM-SHA-1", payload: BinData(0, "biwsbj11c2VyLHI9ZnlrbytkMmxiYkZnT05Sdjlxa3hkYXdM"), options: {skipEmptyExchange: true}}`\
+S:
```

Comment:
Optional: Consider using the new `CMD =`, `RESP =` format for describing the server conversation.

Reply:
Fixed